File: two_pointers/three_sum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        prev = None
        for i in range(len(nums) - 1):
            if nums[i] == prev:
                continue;
            l, r = i + 1, len(nums) - 1
            while l < r:
                one, two, three = nums[i], nums[l], nums[r]
                if(one + two + three > 0):
                    r = r - 1
                elif(one + two + three < 0):
                    l = l + 1
                else:
                    res.append([one, two, three])
                    l = l + 1
                    while nums[l] == nums[l-1] and l < r:
                        l += 1
            prev = nums[i]
        return res

'''
while nums[l] == nums[l-1] and l < r:
  l += 1
This bit is imporant - as we are just forwarding the left pointer to rule out duplicacy.
'''

File: old/two_pointers/three_sum.js
var threeSum = function(nums) {
  const sortedNums = nums.sort((a,b) => a-b);
  console.log(sortedNums);
    const result = [];
    for(let i=0; i<sortedNums.length; i++) {
        if(i>0 && sortedNums[i] === sortedNums[i-1]) {
            continue;
        }
        let left = i+1;
        let right = sortedNums.length - 1;
        while(left < right) {
          let threeSum = sortedNums[left] + sortedNums[right] + sortedNums[i];
          if(threeSum > 0) right--;
          else if(threeSum < 0) left++;
          else {
              result.push([sortedNums[i], sortedNums[left], sortedNums[right]]);
              left++;
              while(sortedNums[left] === sortedNums[left-1] && left<right) {
                  left++;
              }
          }
        }
      }
  return result;
};

console.log(threeSum([-1,0,1,2,-1,-4,-2,-3,3,0,4]))